### PR TITLE
Fix parallel map on OSX for PY38

### LINKED ISSUE #4401: parallel map gives BrokenProcessPool error on OSX in Py38
### Information

- **Qiskit Terra version**: master
- **Python version**: Py3.8.2
- **Operating system**: OSX

### What is the current behavior?

```python
import numpy as np
from qiskit.tools.parallel import parallel_map

def f(x):
    return np.sin(x)/2

for _ in range(10):
    parallel_map(f, np.arange(50))
```

```
BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
```

```
Traceback (most recent call last):
  File "/opt/miniconda3/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/opt/miniconda3/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/miniconda3/lib/python3.8/concurrent/futures/process.py", line 233, in _process_worker
    call_item = call_queue.get(block=True)
  File "/opt/miniconda3/lib/python3.8/multiprocessing/queues.py", line 116, in get
    return _ForkingPickler.loads(res)
AttributeError: Can't get attribute 'f' on <module '__main__' (built-in)>
```

### PR DESCRIPTION
**Pull request by nonhermitian**
### Summary
fixes #4401 

Fix parallel map issues by more or less reverting #3491, and going back to old method.  This now works for me.

### CODE REVIEW

File: qiskit/tools/parallel.py [hunk 1]
```diff
@@ -51,12 +51,18 @@
...
+import sys
 import platform
-from concurrent.futures import ProcessPoolExecutor
+import multiprocessing
 from qiskit.exceptions import QiskitError
 from qiskit.util import local_hardware_info
 from qiskit.tools.events.pubsub import Publisher
 
+if sys.platform == 'darwin':
+    POOL = multiprocessing.get_context('fork').Pool
```
Comment (mtreinish) [inline qiskit/tools/parallel.py:62]: Wasn't there an issue with fork on macOS which is why they changed the default behavior in 3.8 (which prompted #3491 in the first place)?

**Comment by nonhermitian** [comment]
Yeah, I thought the same but it must have fixed upstream somewhere as works fine on 3.8.2.

**Comment by ajavadia** [comment]
I ran this branch and I think it is slowing things down significantly somehow. Maybe it is preventing something from running in parallel?

**Comment by nonhermitian** [comment]
The activity monitor shows that all CPU's being utilized.  When running this new code on Py37 it matches the runtime of the old version.  However, on Py38 I do see a slow down of around 50% compared to the Py37 timing.  However this seems to be a Py38 issue, not the code itself.

**Comment by 1ucian0** [comment]
ping @ajavadia ?

**Comment by ajavadia** [comment]
I just checked this again. I don't think this is the right solution because it creates a massive slowdown. I am on Python 3.7.3, not 3.8.

The tiny example I put above gets about 50% slower. But I have other examples where it just does not terminate. For example try this notebook (start from top and go down until the VQC algorithm section): https://github.com/ajavadia/qiskit-terra/blob/Demo/demo/Circuit%20Library%20Demo.ipynb

I don't know what the problem is. I see that my terminal where Jupyter is running prints many errors of this form:
`libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: Couldn't close file`

File: qiskit/tools/parallel.py [hunk 2]
```diff
@@ -117,12 +123,18 @@ def _callback(_):
        and os.getenv('QISKIT_IN_PARALLEL') == 'FALSE':
         os.environ['QISKIT_IN_PARALLEL'] = 'TRUE'
         try:
-            results = []
-            with ProcessPoolExecutor(max_workers=num_processes) as executor:
-                param = map(lambda value: (task, value, task_args, task_kwargs), values)
-                future = executor.map(_task_wrapper, param)
+            pool = POOL(processes=CPU_COUNT)
```
Comment (mtreinish) [inline qiskit/tools/parallel.py:126]:
```suggestion
            pool = POOL(processes=num_processes)
```

**Comment by ajavadia** [comment]
It appears that calling matplotlib exacerbates the situation. Here's a more minimal example. I'm on Python 3.7.7 and MacOS Catalina.

* If Part 1 and Part 2 are both in the same Jupyter cell, they execute fine. However about 50% slower on this branch vs. master.

* If I split Part 1 and Part 2 between two different Jupyter cells, Part 2 does not terminate with this branch.

It may be an Aqua issue too, as I couldn't make a more minimal Part 2 like this with normal circuit simulation.

```python
# Part 1
import matplotlib.pyplot as plt
fig, ax = plt.subplots(1, 1)

# Part 2
import numpy as np
from qiskit.providers.aer import QasmSimulator
from qiskit.circuit.library import EfficientSU2
from qiskit.circuit.library import ZZFeatureMap
from qiskit.aqua.algorithms import VQC
from qiskit.aqua.components.optimizers import SLSQP

training_data = {'A': np.asarray([[2.95309709, 2.51327412], [3.14159265, 4.08407045]]),
                 'B': np.asarray([[4.08407045, 2.26194671], [4.46106157, 2.38761042]])}
testing_data = {'A': np.asarray([[3.83274304, 2.45044227]]),
                'B': np.asarray([[3.89557489, 0.31415927]])}
feature_dimension = 2

data_preparation = ZZFeatureMap(feature_dimension=feature_dimension, reps=2)
classifier = EfficientSU2(num_qubits=data_preparation.num_qubits, reps=2)
optimizer = SLSQP()

vqc = VQC(optimizer, data_preparation, classifier, training_data, testing_data)
result = vqc.run(QasmSimulator(), seed_simulator=2, seed_transpiler=2)
```

**Comment by nonhermitian** [comment]
I am seeing ~3x slow down on the Aqua example.

**Comment by nonhermitian** [comment]
I am also seeing a 33% slow down on large parallel QV runs.

However, barring the fix here, I am not sure how to proceed with fixing parallel on 38 for OSX.  We could try something like: https://joblib.readthedocs.io/en/latest/parallel.html but I have not tried it.

**Comment by mtreinish** [comment]
Could we special case the python 3.8 execute path? Like check if it's darwin and 3.8 and run things the old way. We take the performance hit, but a necessary one for 3.8. It's not like spawn with 3.8 on osx is fast today, for example see: https://github.com/Qiskit/qiskit-aer/pull/554 and if this makes it more reliable for 3.8 I'm fine with special casing it. It's not like we have a lot of control over this

**Comment by nonhermitian** [comment]
joblib is on par with this PR, so slower than current,

**Comment by nonhermitian** [comment]
Actually a week after I made this (May 13 exactly), Py 3.8.3 came out and the issue seems to be resolved.  Some OSX specific fixes for multiprocessing in there:

>on macOS, the spawn start method is now used by default in multiprocessing

>multiprocessing can now use shared memory segments to avoid pickling costs between processes

Note that I still see slow down on 38 vs 37.  Closing as no longer needed.

**Comment by Cryoris** [comment]
@nonhermitian we're still observing this issue with Python 3.8.3 on MacOS Catalina, running @ajavadia's script slightly modified as:
```python
from qiskit.aqua.components.optimizers import SLSQP
from qiskit.aqua.algorithms import VQC
from qiskit.circuit.library import ZZFeatureMap
from qiskit.circuit.library import EfficientSU2
from qiskit import BasicAer
import numpy as np

# uncommenting the following line works
# if __name__ == '__main__':  
print('test')

training_data = {'A': np.asarray([[2.95309709, 2.51327412], [3.14159265, 4.08407045]]),
                 'B': np.asarray([[4.08407045, 2.26194671], [4.46106157, 2.38761042]])}
testing_data = {'A': np.asarray([[3.83274304, 2.45044227]]),
                'B': np.asarray([[3.89557489, 0.31415927]])}
feature_dimension = 2

data_preparation = ZZFeatureMap(feature_dimension=feature_dimension, reps=2)
classifier = EfficientSU2(num_qubits=data_preparation.num_qubits, reps=2)
optimizer = SLSQP()

vqc = VQC(optimizer, data_preparation, classifier, training_data, testing_data)
result = vqc.run(BasicAer.get_backend('qasm_simulator'), seed_simulator=2, seed_transpiler=2)
```
prints `'test'` several times on 3.8.3, which is not happening for 3.7 and 3.6. Putting the script into a `if __name__ == '__main__'` shield does work, however. 

@nonhermitian can you confirm this? Do you think your fix in this PR could resolve this issue?

**Pull request closed without merging**